core/src/main/java/org/infinispan/marshall/persistence/impl/PersistenceMarshallerImpl.java

```diff
@@ -0,0 +1,234 @@
+   }
+
+   @Override
+   public void register(SerializationContextInitializer initializer) {
+      try {
+         initializer.registerSchema(serializationContext);
+         initializer.registerMarshallers(serializationContext);
+      } catch (IOException e) {
+         throw new CacheException("Exception encountered when initialising the PersistenceMarshaller SerializationContext", e);
+      }
+   }
+
+   @Override
+   public void stop() {
+      userMarshaller.stop();
+   }
+
+   @Override
+   public MediaType mediaType() {
+      return MediaType.APPLICATION_PROTOSTREAM;
+   }
+
+   @Override
+   public ByteBuffer objectToBuffer(Object o) {
+      return objectToBuffer(o, -1);
+   }
+
+   @Override
+   public byte[] objectToByteBuffer(Object obj, int estimatedSize) throws IOException, InterruptedException {
+      ByteBuffer b = objectToBuffer(obj, estimatedSize);
+      byte[] bytes = new byte[b.getLength()];
+      System.arraycopy(b.getBuf(), b.getOffset(), bytes, 0, b.getLength());
+      return bytes;
+   }
+
+   @Override
+   public byte[] objectToByteBuffer(Object obj) throws IOException, InterruptedException {
+      return objectToByteBuffer(obj, sizeEstimate(obj));
+   }
+
+   private ByteBuffer objectToBuffer(Object o, int estimatedSize) {
+      if (o == null)
+         return null;
+
+      try {
+         boolean requiresWrapping = !isPersistenceClass(o);
+         if (requiresWrapping)
+            o = wrapUserObject(o);
+         int size = estimatedSize < 0 ? sizeEstimate(o, true) : estimatedSize;
+         ByteArrayOutputStream baos = new ByteArrayOutputStream(size);
+         ProtobufUtil.toWrappedStream(serializationContext, baos, o, size);
+         byte[] bytes = baos.toByteArray();
+         return new ByteBufferImpl(bytes, 0, bytes.length);
+      } catch (Throwable t) {
+         log.warnf(t, "Cannot marshall %s", o.getClass().getName());
+         throw new MarshallingException(t.getMessage(), t.getCause());
+      }
+   }
+
+   @Override
+   public Object objectFromByteBuffer(byte[] buf) throws IOException, ClassNotFoundException {
+      return objectFromByteBuffer(buf, 0, buf.length);
+   }
+
+   @Override
+   public Object objectFromByteBuffer(byte[] buf, int offset, int length) throws IOException {
+      return unwrapAndInit(ProtobufUtil.fromWrappedByteArray(serializationContext, buf, offset, length));
+   }
+
+   @Override
+   public BufferSizePredictor getBufferSizePredictor(Object o) {
+      // TODO if persistenceClass, i.e. protobuf based, return estimate based upon schema
+      return userMarshaller.getBufferSizePredictor(o);
+   }
+
+   @Override
+   public void writeObject(Object o, OutputStream out) throws IOException {
+      boolean requiresWrapping = !isPersistenceClass(o);
+      if (requiresWrapping)
+         o = wrapUserObject(o);
+      int size = sizeEstimate(o, true);
+      ProtobufUtil.toWrappedStream(serializationContext, out, o, size);
```

Comment:
I really like the way protostream has the stream parameter before the object, not sure how we ended up with the wrong order in our interfaces ;)